¡Hola Ian! Como te va?

Mi nombre es Facundo Lozano! Un gusto conocerte, seré tu revisor en este proyecto.

A continuación un poco sobre la modalidad de revisión que usaremos:

Cuando enccuentro un error por primera vez, simplemente lo señalaré, te dejaré encontrarlo y arreglarlo tú cuenta. Además, a lo largo del texto iré haciendo algunas observaciones sobre mejora en tu código y también haré comentarios sobre tus percepciones sobre el tema. Pero si aún no puedes realizar esta tarea, te daré una pista más precisa en la próxima iteración y también algunos ejemplos prácticos. Estaré abierto a comentarios y discusiones sobre el tema.

Encontrará mis comentarios a continuación: **no los mueva, modifique ni elimine**.

Puedes encontrar mis comentarios en cuadros verdes, amarillos o rojos como este:

<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Exito. Todo se ha hecho de forma exitosa.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Observación. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">

<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Necesita arreglos. Este apartado necesita algunas correcciones. El trabajo no puede ser aceptado con comentarios rojos. 
</div>

Puedes responder utilizando esto:

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante.</b> <a class="tocSkip"></a>
</div>

# Preciador de vehículos

<div class="alert alert-block alert-success">
<b>Review General. (Iteración 1) </b> <a class="tocSkip"></a>

Ian. antes que nada disculpas por las demoras. Se ve que hubo una confusión del revisor anterior por lo que a partir de ahora comenzaré esta iteración yo. Un gusto!    

Siempre me tomo este tiempo al inicio de tu proyecto para comentarte mis apreciaciones generales de esta iteración de tu entrega.

Me gusta comenzar dando la bienvenida al mundo de los datos a los estudiantes, te deseo lo mejor y espero que consigas lograr tus objetivos. Personalmente me gusta brindar el siguiente consejo, "Está bien equivocarse, es normal y es lo mejor que te puede pasar. Aprendemos de los errores y eso te hará mejor programandor ya que podrás descubrir cosas a medida que avances y son estas cosas las que te darán esa experiencia para ser un gran Data Scientist"
    
Ahora si yendo a esta notebook. Felicitaciones Ian porque has hehco un muy buen trabajo a lo largo de todo el proyecto, se nota tu gran manejo de python y de las herramientas, a la vez te destacaste por el orden y el clean code, bien hecho! Simplemente hubo algunos detalles que deberíamos corregir pero no te preocupes que seguro te tomará unos momentos, te he dejado unos comentarios rojos con el contexto necesario para que lo resuelvas.
    
    
Éxitos y espero con ansias a nuestra próxima iteración. Saludos Ian!

El servicio de venta de autos usados Rusty Bargain está desarrollando una aplicación para atraer nuevos clientes. Gracias a esa app, puedes averiguar rápidamente el valor de mercado de tu coche. Tienes acceso al historial: especificaciones técnicas, versiones de equipamiento y precios. Tienes que crear un modelo que determine el valor de mercado.
A Rusty Bargain le interesa:
- la calidad de la predicción;
- la velocidad de la predicción;
- el tiempo requerido para el entrenamiento

## Preparación de datos

### Inicialización

In [2]:
import numpy as np
import pandas as pd
import lightgbm as lgb

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeRegressor
from lightgbm import LGBMRegressor
from sklearn.metrics import mean_squared_error
from math import sqrt

### Carga de datos

In [3]:
try:
    df = pd.read_csv('/datasets/car_data.csv')
except:
    df = pd.read_csv('datasets/car_data.csv')

### Exploración de datos

In [4]:
df.head()

DateCrawled  Price VehicleType  RegistrationYear Gearbox  Power  \
0  24/03/2016 11:52    480         NaN              1993  manual      0   
1  24/03/2016 10:58  18300       coupe              2011  manual    190   
2  14/03/2016 12:52   9800         suv              2004    auto    163   
3  17/03/2016 16:54   1500       small              2001  manual     75   
4  31/03/2016 17:25   3600       small              2008  manual     69   

   Model  Mileage  RegistrationMonth  FuelType       Brand NotRepaired  \
0   golf   150000                  0    petrol  volkswagen         NaN   
1    NaN   125000                  5  gasoline        audi         yes   
2  grand   125000                  8  gasoline        jeep         NaN   
3   golf   150000                  6    petrol  volkswagen          no   
4  fabia    90000                  7  gasoline       skoda          no   

        DateCreated  NumberOfPictures  PostalCode          LastSeen  
0  24/03/2016 00:00                 0       70435  07/04/2016 03:16  
1  24/03/2016 00:00                 0       66954  07/04/2016 01:46  
2  14/03/2016 00:00                 0       90480  05/04/2016 12:47  
3  17/03/2016 00:00                 0       91074  17/03/2016 17:40  
4  31/03/2016 00:00                 0       60437  06/04/2016 10:17

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 354369 entries, 0 to 354368
Data columns (total 16 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   DateCrawled        354369 non-null  object
 1   Price              354369 non-null  int64 
 2   VehicleType        316879 non-null  object
 3   RegistrationYear   354369 non-null  int64 
 4   Gearbox            334536 non-null  object
 5   Power              354369 non-null  int64 
 6   Model              334664 non-null  object
 7   Mileage            354369 non-null  int64 
 8   RegistrationMonth  354369 non-null  int64 
 9   FuelType           321474 non-null  object
 10  Brand              354369 non-null  object
 11  NotRepaired        283215 non-null  object
 12  DateCreated        354369 non-null  object
 13  NumberOfPictures   354369 non-null  int64 
 14  PostalCode         354369 non-null  int64 
 15  LastSeen           354369 non-null  object
dtypes: int64(7), object(

In [6]:
df.describe()

Price  RegistrationYear          Power        Mileage  \
count  354369.000000     354369.000000  354369.000000  354369.000000   
mean     4416.656776       2004.234448     110.094337  128211.172535   
std      4514.158514         90.227958     189.850405   37905.341530   
min         0.000000       1000.000000       0.000000    5000.000000   
25%      1050.000000       1999.000000      69.000000  125000.000000   
50%      2700.000000       2003.000000     105.000000  150000.000000   
75%      6400.000000       2008.000000     143.000000  150000.000000   
max     20000.000000       9999.000000   20000.000000  150000.000000   

       RegistrationMonth  NumberOfPictures     PostalCode  
count      354369.000000          354369.0  354369.000000  
mean            5.714645               0.0   50508.689087  
std             3.726421               0.0   25783.096248  
min             0.000000               0.0    1067.000000  
25%             3.000000               0.0   30165.000000  
50%             6.000000               0.0   49413.000000  
75%             9.000000               0.0   71083.000000  
max            12.000000               0.0   99998.000000

<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>
    
Excelente Ian, hasta el momento hemos cargado correctamente los datos implementando un bloque **try and except** para mitigar posibles errores, a la vez has implementado métodos para visualizar y comprender la composición de nuestros datos. Bien hecho!

Se cambian los valores no y yes de la columna NotRepaired por valores 0 y 1 para poderlo usar para el entrenamiento del modelo.

In [7]:
df['NotRepaired'] = df['NotRepaired'].replace('no',0)
df['NotRepaired'] = df['NotRepaired'].replace('yes',1)
df['NotRepaired'] = df['NotRepaired'].fillna(0)

C:\Users\reiva\AppData\Local\Temp\ipykernel_5384\1297644283.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['NotRepaired'] = df['NotRepaired'].replace('yes',1)


<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Una muy buena decisión la de reemplazar los valores de la columna **NotRepaired** y excelente a la hora de completar los nulos.

In [8]:
df.head()

DateCrawled  Price VehicleType  RegistrationYear Gearbox  Power  \
0  24/03/2016 11:52    480         NaN              1993  manual      0   
1  24/03/2016 10:58  18300       coupe              2011  manual    190   
2  14/03/2016 12:52   9800         suv              2004    auto    163   
3  17/03/2016 16:54   1500       small              2001  manual     75   
4  31/03/2016 17:25   3600       small              2008  manual     69   

   Model  Mileage  RegistrationMonth  FuelType       Brand  NotRepaired  \
0   golf   150000                  0    petrol  volkswagen          0.0   
1    NaN   125000                  5  gasoline        audi          1.0   
2  grand   125000                  8  gasoline        jeep          0.0   
3   golf   150000                  6    petrol  volkswagen          0.0   
4  fabia    90000                  7  gasoline       skoda          0.0   

        DateCreated  NumberOfPictures  PostalCode          LastSeen  
0  24/03/2016 00:00                 0       70435  07/04/2016 03:16  
1  24/03/2016 00:00                 0       66954  07/04/2016 01:46  
2  14/03/2016 00:00                 0       90480  05/04/2016 12:47  
3  17/03/2016 00:00                 0       91074  17/03/2016 17:40  
4  31/03/2016 00:00                 0       60437  06/04/2016 10:17

Verificamos la cantidad de valores ausentes.

In [9]:
na_ratio = ((df.isnull().sum() / len(df))*100).sort_values(ascending = False)
na_ratio

VehicleType          10.579368
FuelType              9.282697
Gearbox               5.596709
Model                 5.560588
DateCrawled           0.000000
Price                 0.000000
RegistrationYear      0.000000
Power                 0.000000
Mileage               0.000000
RegistrationMonth     0.000000
Brand                 0.000000
NotRepaired           0.000000
DateCreated           0.000000
NumberOfPictures      0.000000
PostalCode            0.000000
LastSeen              0.000000
dtype: float64

Observamos que todas las columnas donde hay valores ausentes son columnas con valores cualitativos que no afectan el entrenamiento del modelo, por tanto, no se imputan los valores ausentes.

Verificamos la cantidad de valores duplicados.

In [10]:
df.duplicated().sum()

291

Encontramos valores duplicados y para no alterar el entrenamiento del modelo se decide eliminar los valores duplicados. Adicionalmente, son muy pocos valores ausentes comparados con el total de filas del dataframe.

In [11]:
df = df.drop_duplicates()

In [14]:
df['RegistrationYear'].unique()

array([1993, 2011, 2004, 2001, 2008, 1995, 1980, 2014, 1998, 2005, 1910,
       2016, 2007, 2009, 2002, 2018, 1997, 1990, 2017, 1981, 2003, 1994,
       1991, 1984, 2006, 1999, 2012, 2010, 2000, 1992, 2013, 1996, 1985,
       1989, 2015, 1982, 1976, 1983, 1973, 1111, 1969, 1971, 1987, 1986,
       1988, 1970, 1965, 1945, 1925, 1974, 1979, 1955, 1978, 1972, 1968,
       1977, 1961, 1960, 1966, 1975, 1963, 1964, 5000, 1954, 1958, 1967,
       1959, 9999, 1956, 3200, 1000, 1941, 8888, 1500, 2200, 4100, 1962,
       1929, 1957, 1940, 3000, 2066, 1949, 2019, 1937, 1951, 1800, 1953,
       1234, 8000, 5300, 9000, 2900, 6000, 5900, 5911, 1933, 1400, 1950,
       4000, 1948, 1952, 1200, 8500, 1932, 1255, 3700, 3800, 4800, 1942,
       7000, 1935, 1936, 6500, 1923, 2290, 2500, 1930, 1001, 9450, 1944,
       1943, 1934, 1938, 1688, 2800, 1253, 1928, 1919, 5555, 5600, 1600,
       2222, 1039, 9996, 1300, 8455, 1931, 1915, 4500, 1920, 1602, 7800,
       9229, 1947, 1927, 7100, 8200, 1946, 7500, 35

In [15]:
df = df[df['RegistrationYear'] <= 2024]
df['RegistrationYear'].unique()

array([1993, 2011, 2004, 2001, 2008, 1995, 1980, 2014, 1998, 2005, 1910,
       2016, 2007, 2009, 2002, 2018, 1997, 1990, 2017, 1981, 2003, 1994,
       1991, 1984, 2006, 1999, 2012, 2010, 2000, 1992, 2013, 1996, 1985,
       1989, 2015, 1982, 1976, 1983, 1973, 1111, 1969, 1971, 1987, 1986,
       1988, 1970, 1965, 1945, 1925, 1974, 1979, 1955, 1978, 1972, 1968,
       1977, 1961, 1960, 1966, 1975, 1963, 1964, 1954, 1958, 1967, 1959,
       1956, 1000, 1941, 1500, 1962, 1929, 1957, 1940, 1949, 2019, 1937,
       1951, 1800, 1953, 1234, 1933, 1400, 1950, 1948, 1952, 1200, 1932,
       1255, 1942, 1935, 1936, 1923, 1930, 1001, 1944, 1943, 1934, 1938,
       1688, 1253, 1928, 1919, 1600, 1039, 1300, 1931, 1915, 1920, 1602,
       1947, 1927, 1946], dtype=int64)

In [16]:
df.reset_index(inplace=True)

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 353973 entries, 0 to 353972
Data columns (total 18 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   level_0            353973 non-null  int64  
 1   index              353973 non-null  int64  
 2   DateCrawled        353973 non-null  object 
 3   Price              353973 non-null  int64  
 4   VehicleType        316594 non-null  object 
 5   RegistrationYear   353973 non-null  int64  
 6   Gearbox            334224 non-null  object 
 7   Power              353973 non-null  int64  
 8   Model              334314 non-null  object 
 9   Mileage            353973 non-null  int64  
 10  RegistrationMonth  353973 non-null  int64  
 11  FuelType           321160 non-null  object 
 12  Brand              353973 non-null  object 
 13  NotRepaired        353973 non-null  float64
 14  DateCreated        353973 non-null  object 
 15  NumberOfPictures   353973 non-null  int64  
 16  Po

<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Nuevamente muy bien hecho, en este caso hemos cumplido con observar  la cantidad de nulos en nuestras columnas y los duplicados, bien hecho! Una recomendación siempre luego de eliminar nulos es que apliquemos el método **reset_index()**, te invito a ivnestigarlo!

<div class="alert alert-block alert-warning">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Aquí podríamos implementar más cambios antes de pasar a nuestra próxima sección Ian, deberíamos observar las columnas y sus valores. Por ejemplo, el **RegistrationYear** tiene valores que no son lógicos como por ejemplo 2050, esos casos podríamos eliminarlos por ejemplo. Te invito a investigar posibles correcciones en otras columnas.

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante. (Iteración 1)</b> <a class="tocSkip"></a>

Se agregó el método reset_index y se eliminaron los valores de **RegistrationYear** superiores a 2024.
</div>

## Entrenamiento del modelo 

Seleccionamos las columnas que conformarán las características y el objetivo. Para realizar esto primero usaremos OHE para convertir las columnas que aporten información de cualitativas a cuantitativas.

#### Codificación OHE

In [19]:
print(df['Gearbox'].unique())
print(df['FuelType'].unique())

['manual' 'auto' nan]
['petrol' 'gasoline' nan 'lpg' 'other' 'hybrid' 'cng' 'electric']


In [20]:
df_ohe = pd.get_dummies(df, columns=['Gearbox', 'FuelType'])
df_ohe

level_0   index       DateCrawled  Price  VehicleType  \
0             0       0  24/03/2016 11:52    480          NaN   
1             1       1  24/03/2016 10:58  18300        coupe   
2             2       2  14/03/2016 12:52   9800          suv   
3             3       3  17/03/2016 16:54   1500        small   
4             4       4  31/03/2016 17:25   3600        small   
...         ...     ...               ...    ...          ...   
353968   354073  354364  21/03/2016 09:50      0          NaN   
353969   354074  354365  14/03/2016 17:48   2200          NaN   
353970   354075  354366  05/03/2016 19:56   1199  convertible   
353971   354076  354367  19/03/2016 18:57   9200          bus   
353972   354077  354368  20/03/2016 19:41   3400        wagon   

        RegistrationYear  Power        Model  Mileage  RegistrationMonth  ...  \
0                   1993      0         golf   150000                  0  ...   
1                   2011    190          NaN   125000                  5  ...   
2                   2004    163        grand   125000                  8  ...   
3                   2001     75         golf   150000                  6  ...   
4                   2008     69        fabia    90000                  7  ...   
...                  ...    ...          ...      ...                ...  ...   
353968              2005      0         colt   150000                  7  ...   
353969              2005      0          NaN    20000                  1  ...   
353970              2000    101       fortwo   125000                  3  ...   
353971              1996    102  transporter   150000                  3  ...   
353972              2002    100         golf   150000                  6  ...   

                LastSeen  Gearbox_auto Gearbox_manual  FuelType_cng  \
0       07/04/2016 03:16         False           True         False   
1       07/04/2016 01:46         False           True         False   
2       05/04/2016 12:47          True          False         False   
3       17/03/2016 17:40         False           True         False   
4       06/04/2016 10:17         False           True         False   
...                  ...           ...            ...           ...   
353968  21/03/2016 10:42         False           True         False   
353969  06/04/2016 00:46         False          False         False   
353970  11/03/2016 18:17          True          False         False   
353971  07/04/2016 07:15         False           True         False   
353972  24/03/2016 12:45         False           True         False   

        FuelType_electric FuelType_gasoline  FuelType_hybrid  FuelType_lpg  \
0                   False             False            False         False   
1                   False              True            False         False   
2                   False              True            False         False   
3                   False             False            False         False   
4                   False              True            False         False   
...                   ...               ...              ...           ...   
353968              False             False            False         False   
353969              False             False            False         False   
353970              False             False            False         False   
353971              False              True            False         False   
353972              False              True            False         False   

        FuelType_other  FuelType_petrol  
0                False             True  
1                False            False  
2                False            False  
3                False             True  
4                False            False  
...                ...              ...  
353968           False             True  
353969           False            False  
353970           False             True  
353971           False            False  
35

In [21]:
features = df_ohe[['RegistrationYear','Power','Mileage','NotRepaired', 'Gearbox_auto', 'Gearbox_manual', 'FuelType_cng', 'FuelType_electric', 'FuelType_gasoline', 'FuelType_hybrid', 'FuelType_lpg', 'FuelType_other', 'FuelType_petrol']]
target = df['Price']

<div class="alert alert-block alert-danger">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Cuidado aquí Ian, existen columnas que si nos aportan información pero que no estamos utilizando. Por un lado te felicito porque con este filtrado de features si nos estamos quitando columnas que no nos aportan sin embargo también estamos quitando las que si. Ten en cuenta que a todas aquellas columnas categóricas debemos utilizarlas también, debes tratar de aplicar un método de **encoding** para pasar estas variables de categóricas a numéricas.

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante. (Iteración 1)</b> <a class="tocSkip"></a>

Se agregó codifcacion OHE a las columnas **Gearbox** y **Fueltype** para usarlas en el modelo.
</div>

Dividimos el dataframe en entrenamiento, test y validación en una proporción 

In [22]:
features_train, features_test, target_train, target_test = train_test_split(features, target, test_size=0.2, random_state=42)
features_train, features_val, target_train, target_val = train_test_split(features_train, target_train, test_size=0.25, random_state=42)

<div class="alert alert-block alert-danger">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Recuerda que debemos dividir nuestros datos en 3 conjuntos que son entrenamiento, validación y testeo. Deberíamos aplicar nuevamente el método **train_test_split**, investiga como dividir en estos conjuntos. Donde en nuestros modelos debemos entrenar y validar con el conjunto de entrenamiento y validación, luego al final deberíamos elegir el mejor modelo y testearlo sobre el conjunto de prueba.

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante. (Iteración 1)</b> <a class="tocSkip"></a>

Se dividé el dataframe en 3 conjuntos **(train, validation y test)**
</div>

### Regresión lineal

In [23]:
%%time

lr = LinearRegression()
lr.fit(features_train, target_train)
pred_lr = lr.predict(features_val)

CPU times: total: 109 ms
Wall time: 183 ms


<div class="alert alert-block alert-warning">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Un modelo tradicional y útil, bien hecho Ian. Recuerda que en este caso deberíamos entrenar con tran y testear sobre el **valid**.

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante. (Iteración 1)</b> <a class="tocSkip"></a>

Se testea sobre **valid**
</div>

### Árboles de decisión

In [24]:
%%time
model = DecisionTreeRegressor(random_state=42)
param_grid = {
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 10, 20],
    'min_samples_leaf': [1, 5, 10]
}

grid_search = GridSearchCV(model, param_grid, cv=5, scoring='neg_mean_squared_error')
grid_search.fit(features_train, target_train)

best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

pred_dt = best_model.predict(features_val)

CPU times: total: 45.9 s
Wall time: 1min 36s


<div class="alert alert-block alert-warning">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Nuevamente bien hecho Ian! Has implementado otro modelo y en este caso creando un gridsearch para la optimización en busqueda de hiperparámetros. Bien hecho! Recuerda modificar los conjuntos de datos que utilizamos.
    

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante. (Iteración 1)</b> <a class="tocSkip"></a>

Se testea sobre **valid**
</div>

### LightGBM

In [25]:
%%time

lgbm = LGBMRegressor(random_state=42)

# Definimos la grilla de hiperparámetros a probar
param_grid_lgbm = {
    'n_estimators': [100, 200],
    'learning_rate': [0.1, 0.01],
}

grid_search_lgbm = GridSearchCV(lgbm, param_grid_lgbm, cv=3, scoring='neg_mean_squared_error')
grid_search_lgbm.fit(features_train, target_train)

pred_lgbm = grid_search_lgbm.predict(features_val)

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.005389 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 372
[LightGBM] [Info] Number of data points in the train set: 141588, number of used features: 13
[LightGBM] [Info] Start training from score 4425.915183
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.004253 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 374
[LightGBM] [Info] Number of data points in the train set: 141589, number of used features: 13
[LightGBM] [Info] Start training from score 4428.042214
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.005082 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.004566 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 372
[LightGBM] [Info] Number of data points in the train set: 141588, number of used features: 13
[LightGBM] [Info] Start training from score 4425.915183
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.005946 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 374
[LightGBM] [Info] Number of data points in the train set: 141589, number of used features: 13
[LightGBM] [Info] Start training from score 4428.042214
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.006015 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not

<div class="alert alert-block alert-warning">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

    
Excelente! Un tercer modelo y nuevamente buscando mejora de hiperparámetros. Bien hecho Ian! Deberíamos únicamente modificar los conjuntos testeando sobre el conjunto de validación.

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante. (Iteración 1)</b> <a class="tocSkip"></a>

Se testea sobre **valid**
</div>

## Análisis del modelo

### Regresión lineal

In [26]:
rmse_lr = mean_squared_error(target_test, pred_lr,squared=False)
print(f'RMSE para Regresión Lineal: {rmse_lr}')

RMSE para Regresión Lineal: 5229.4752797032315


c:\Users\reiva\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\metrics\_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


### Árboles de decisión

In [27]:
rmse_dt = mean_squared_error(target_test, pred_dt, squared=False)
print(f'RMSE para Bosque Aleatorio: {rmse_dt}')

RMSE para Bosque Aleatorio: 6051.6335517143425


c:\Users\reiva\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\metrics\_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


### LightGBM

In [28]:
rmse_lgbm = mean_squared_error(target_test, pred_lgbm, squared=False)
print(f'RMSE para LightGBM: {rmse_lgbm}')

RMSE para LightGBM: 5981.362174112902


c:\Users\reiva\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\metrics\_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


El modelo de regresión lineal es el que toma menos tiempo entrenar con un resultado de tiempo de CPU de 109 ms, de ahí le sigue LightGBM con un tiempo de CPU de 26.1 s y finalmente, árboles de deicsión con un tiempo de CPU de 45.9 s. La diferencia es muy grande entre los diferentes modelos, pero donde el modelo LighGBM destaca es en su calidad, ya que nos arroja un resultado mejor de 5981.36. Hasta este punto, el modelo LightGBM parece no superar al modelo de regresión lineal, pero procederemos a probar ambos en el conjunto de test para comprobarlo.

<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Excelente muestreo de resultados de los modelos!  A la vez has mostrados los tiempos correctamente en los momentos de entrenar.

<div class="alert alert-block alert-danger">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Antes de dar nuestra conclusión deberíamos agregar una sección Ian, hasta ahora nosotros hemos entrenado y evaluado los modelos, como te he dicho antes deberíamos haber creado 3 conjuntos de datos, uno de ellos (el que falto) era el de validación, este es el que se debería haber usado en todos los testeos hata ahora. En base a los resultados que obtendríamos en estas verificaciones de arriba deberíamos agarrar al modelo que menor **rmse** dio y a este testearlo sobre el tercer conjunto que sería el separado al inicio para el testeo final, unos datos que nuestro modelo nunca haya visto. En resumen, deberías agregar esta sección donde testear al modelo que mejor resultados nos dio aquí arriba con el conjunto de prueba.

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante. (Iteración 1)</b> <a class="tocSkip"></a>

Se agrega la sección mencionada.
</div>

### Evaluación del mejor modelo

In [30]:
pred_lr = lr.predict(features_val)
rmse_lr = mean_squared_error(target_test, pred_lr,squared=False)
print(f'RMSE para Regresión Lineal: {rmse_lr}')

RMSE para Regresión Lineal: 5229.4752797032315


c:\Users\reiva\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\metrics\_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


In [29]:
pred_lgbm = grid_search_lgbm.predict(features_test)
rmse_lgbm = mean_squared_error(target_test, pred_lgbm, squared=False)
print(f'RMSE para LightGBM: {rmse_lgbm}')

RMSE para LightGBM: 2071.9561205530636


c:\Users\reiva\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\metrics\_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


## Conlusiones

El modelo de regresión lineal es el que toma menos tiempo entrenar y hasta el punto donde utilizamos los conjuntos de entrenamiento y validación parecía ser el más efectivo. Pero al utilizar el conjunto de test observamos que el desemmpeño del modelo LightGBM es superior y por lo tanto, el modelo más recomendable a utilizar para esta aplicación.

<div class="alert alert-block alert-warning">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Muy buena conclusión Ian, esto demuestra tu comprensión sobre lo trabajado y obtenido. Bien hecho! Simplemente deberíamos agregas/ajustar esta conclusión luego de corregir lo marcado a lo largo del proyecto.

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante. (Iteración 1)</b> <a class="tocSkip"></a>

Se realizaron todas las modificaciones de las secciones en rojo y en amarillo.
</div>

# Lista de control

Escribe 'x' para verificar. Luego presiona Shift+Enter

- [x]  Jupyter Notebook está abierto
- [x]  El código no tiene errores
- [x]  Las celdas con el código han sido colocadas en orden de ejecución
- [x]  Los datos han sido descargados y preparados
- [x]  Los modelos han sido entrenados
- [x]  Se realizó el análisis de velocidad y calidad de los modelos